src/xds/server.rs

```diff
@@ -0,0 +1,119 @@
+use std::sync::Arc;
+
+use cached::{Cached, CachedAsync};
+use futures::Stream;
+
+use crate::xds::{
+    service::discovery::v3::{
+        aggregated_discovery_service_server::AggregatedDiscoveryService, DeltaDiscoveryRequest,
+        DeltaDiscoveryResponse, DiscoveryRequest, DiscoveryResponse,
+    },
+    Cache, DiscoveryServiceProvider, ResourceType,
+};
+
+pub struct ControlPlane {
+    provider: Arc<dyn DiscoveryServiceProvider>,
+}
+
+impl ControlPlane {
+    /// Creates a new server for a [DiscoveryServiceProvider].
+    pub fn new<P: DiscoveryServiceProvider + 'static>(provider: P) -> Self {
+        Self {
+            provider: Arc::from(provider),
+        }
+    }
+
+    /// Creates a new server from a dynamic reference
+    /// counted [DiscoveryServiceProvider].
+    pub fn from_arc(provider: Arc<dyn DiscoveryServiceProvider>) -> Self {
+        Self { provider }
+    }
+
+    /// Creates a new server for a [DiscoveryServiceProvider] with a cache.
+    pub fn with_cache<P: DiscoveryServiceProvider + 'static>(provider: P) -> Self {
+        Self::new(Cache::new(provider))
+    }
+}
+
+#[tonic::async_trait]
```

Comment:
Does it make sense to pull in `async_trait` crate into Config.toml and use it directly? 

Looking at [the docs](https://docs.rs/tonic/latest/tonic/attr.async_trait.html), it says "A re-export of async-trait for use with codegen." -- small nit, but feels a bit weird to being using it in our own code?? 🤔